text/0000-module-unification.md

```diff
@@ -0,0 +1,837 @@
+to describe a broad swath of UI concerns. Using the same term for the collection
+of template-invocable UI elements will lower the learning curve for developers
+who are new to Ember, while allowing for a useful set of specialized terms to
+flourish to describe particular _types_ of components.
+
+We've already started down the road of component specialization by introducing
+the concept of "routable components". Once we start actually using "routable
+components" in practice, it will become necessary to refer to plain old
+components as something more specific, like "template components". And this
+distinction will probably lead to plain old helpers being referred to as
+"template helpers". Other concepts, such as "Glimmer components" and "template
+component modifiers" will soon be mixed in. We will end up with a multi-faceted
+toolbox available at the template layer which deserves a simple name that
+matches developer expectations. The general term "components" seems a good fit.
+
+### Scope
+
+Developers should understand the available levels of module scope, as well as
+when each is appropriate to use. Scope should be considered when modules are
+generated, and developers should feel free to move modules if they expand or
+contract in scope.
+
+The following levels of scope should be understood:
+
+* Private - private collections should be used when a component or utility
+  function is needed from a single namespace.
+
+* Project - top-level, project-wide collections should be used for modules that
+  are needed throughout a project.
+
+* Local package - namespaced collections can be useful to group a common set of
+  cross-cutting concerns within a project.
+
+* Local engine - a type of local package that encapsulates a set of
+  functionality that benefits from run-time isolation and strict dependency
+  sharing.
+
+### Testing
+
+Unit, integration, and some acceptance tests can now be co-located with their
+associated modules. Co-location should be encouraged because it makes test
+modules easier to locate in the file system, and easier to move if a module's
+scope changes.
+
+Robert Jackson plans to adapt the
+[Grand Testing Unification RFC](https://github.com/emberjs/rfcs/pull/119)
+to illustrate test co-location and to introduce module types for tests.
+
+# Drawbacks
+
+Any change to a pattern as fundamental as file naming will incur some mental
+friction for developers who are accustomed to the current conventions. It is
+hoped that tooling like Robert's migrator and Ember Watson can lessen this
+friction by automating transitions, and that updated guides, generators, and
+blueprints can make these conventions easy to follow.
+
+Of course, we won't prevent usage of the currently used patterns for some time,
+but they will eventually be deprecated. Some efficiencies, especially in the
+resolver, may not be fully realized until the new patterns are used throughout
```

Comment:
@rtm the only potential one I'm aware of in the resolver is that is currently supports classic and pods style. So if the module doesn't exist in one, it attempt to look up in the other. However, at that point it caches the module so additional lookups are not penalized. Perhaps there are additional perf issues I am unaware of however.